lib/services/local/generic.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package local
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// marshalFunc is a type signature for a marshaling function.
+type marshalFunc[T types.Resource] func(T, ...services.MarshalOption) ([]byte, error)
+
+// UnmarshalFunc is a type signature for an unmarshaling function.
+type unmarshalFunc[T types.Resource] func([]byte, ...services.MarshalOption) (T, error)
+
+// genericResourceService is a generic service for interacting with resources in the backend.
+type genericResourceService[T types.Resource] struct {
+	backend       backend.Backend
+	resourceKind  string
+	limit         int
+	backendPrefix string
+	marshalFunc   marshalFunc[T]
+	unmarshalFunc unmarshalFunc[T]
+}
+
+// ListResources returns a paginated list of resources.
+func (s *genericResourceService[T]) listResources(ctx context.Context, pageSize int, pageToken string) ([]T, string, error) {
+	rangeStart := backend.Key(s.backendPrefix, pageToken)
+	rangeEnd := backend.RangeEnd(rangeStart)
```

Comment:
```suggestion
	rangeEnd := backend.RangeEnd(backend.Key(s.backendPrefix, ""))
```

Reply:
@espadolini @mdwn why was this suggested? won't that mess up the range? I came across this line for an unrelated reason and found this discussion :)

Reply:
I think it should be:
```go
	rangeStart := backend.Key(s.backendPrefix, pageToken)
	rangeEnd := backend.RangeEnd(rangeStart)
```

possibly even using `backend.ExactKey`, but I can't tell for sure if that will work here as I don't know if pageToken is a namespace like `<backendprefix>/<pageToken>/<something>`

Reply:
oh wait, nevermind I think this suggestion is correct since `pageToken` just represents where we left off and further items may not have that as part of their key path >.<

Reply:
I really wish we had backend read APIs based on prefixes rather than just ranges :(